bookkeeper-server/src/main/java/org/apache/bookkeeper/client/BookKeeperAdmin.java

```diff
@@ -1430,19 +1432,44 @@ public void triggerAudit()
     public void decommissionBookie(BookieSocketAddress bookieAddress)
             throws CompatibilityException, UnavailableException, KeeperException, InterruptedException, IOException,
             BKAuditException, TimeoutException, BKException {
+        int maxSleepTimeInBetweenChecks = 10 * 60 * 1000; // 10 minutes
+        int sleepTimePerLedger = 10 * 1000; // 10 secs
+
+        decommissionBookie(
+            bookieAddress,
+            // default wait strategy is to sleep for 30 secs.
+            // so that Auditor gets chance to trigger its
+            // force audittask and let the underreplicationmanager process
+            // to do its replication process
+            () -> {
+                Thread.sleep(30 * 1000);
+                return null;
+            },
+            sleepTimePerLedger,
+            maxSleepTimeInBetweenChecks);
+    }
+
+    @VisibleForTesting
+    public void decommissionBookie(BookieSocketAddress bookieAddress,
+                                   Callable<Void> waitAuditorStrategy,
+                                   int sleepTimePerLedger,
+                                   int maxSleepTimeInBetweenChecks)
+            throws CompatibilityException, UnavailableException, KeeperException, InterruptedException, IOException,
+            BKAuditException, TimeoutException, BKException {
         if (getAvailableBookies().contains(bookieAddress) || getReadOnlyBookies().contains(bookieAddress)) {
             LOG.error("Bookie: {} is not shutdown yet", bookieAddress);
             throw BKException.create(BKException.Code.IllegalOpException);
         }
 
         triggerAudit();
 
-        /*
-         * Sleep for 30 secs, so that Auditor gets chance to trigger its
-         * force audittask and let the underreplicationmanager process
-         * to do its replication process
-         */
-        Thread.sleep(30 * 1000);
+        try {
+            waitAuditorStrategy.call();
+        } catch (InterruptedException ie) {
+            throw ie;
+        } catch (Exception e) {
+            throw new TimeoutException("Encountered exception on waiting auditor auditing the bookies");
```

Comment:
not sure if the TimeoutException is the appropriate one in this case.

Reply:
okay any better exceptions?

Reply:
BKAuditException? even in the InterruptedException case also. 

This pattern is used in other places also

https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/Auditor.java#L692